services/httpd/emitter.go

```diff
@@ -0,0 +1,290 @@
+package httpd
+
+import (
+	"io"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/influxdata/influxdb/influxql"
+	"github.com/influxdata/influxdb/models"
+)
+
+type Encoder interface {
+	// ContentType returns the content-type used to identify this format in HTTP.
+	ContentType() string
+
+	// Encode encodes the full response from the results channel.
+	Encode(w io.Writer, header ResponseHeader, results <-chan *influxql.ResultSet)
+
+	// Error encodes a top-level error to the io.Writer.
+	Error(w io.Writer, err error)
+}
+
+type ResponseFormatter interface {
+	WriteResponse(w io.Writer, resp Response) (int, error)
+	WriteError(w io.Writer, err error)
+	ContentType() string
+}
+
+func NewEncoder(r *http.Request, config *Config) Encoder {
+	epoch := strings.TrimSpace(r.FormValue("epoch"))
+	switch r.Header.Get("Accept") {
```

Comment:
The parsing of the `Accept` header needs to be more complex. It's possible for the client to accept multiple types. This will be a very likely occurrence once msgpack is introduced as clients will prefer msgpack, but also accept json.
For example:

```
Accept: application/x-msgpack, application/json;q=0.7
```

It's also technically valid to do:
```
Accept: application/x-msgpack
Accept: application/json
```
Though this is rather uncommon.